platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultAttributeMatcher.java

```diff
@@ -62,28 +62,75 @@ public AttributeSelectionSchema getSelectionSchema() {
     }
 
     @Override
-    public boolean isMatching(AttributeContainerInternal candidate, AttributeContainerInternal requested) {
+    public boolean isMatching(ImmutableAttributes candidate, ImmutableAttributes requested) {
```

Comment:
Using `ImmutableAttributes` here is a matter of asserting the memory safety of the code over any other reason. Its nice implementing this code knowing that nobody will go and mutate any of these containers after the fact. We can even go ahead and cache some of the results here without fear of the container being mutated after the fact. 

Most of dependency management here uses `ImmutableAttributes` (and `ImmutableCapabilities`) throughout so its not a stretch to require them here. Dependency management deals with a lot of immutable metadata types, caching, etc, and its nice to explicitly know you're working with immutable data. 

As for adding `findEntry` to `AttributeContainerInternal`, I wouldn't be against that. Theoretically any attribute container _should_ be able to implement `findEntry`, but `ImmutableAttributes` was designed to be _really fast_ at doing that (just look at the implementation of that method). Other containers would be slow. Also, its more work to implement it for other containers, and why would we if its easy enough to get an instance of `ImmutableAttributes`.

Reply:
As for your second question, I do not think we should switch to `first` and `second` for the first method. 

For the second method (`weaklyMatches`) the ordering of the parameters _does not matter_ since we do attribute matching in both directions. 

However, for the first method, the "ordering" of the parameters is _very_ important. Attribute compatibility and disambiguation rules are _build_ on having a producer and consumer value. Directionality is built in. And those rules use the `producer` and `consumer` terminology (and `candidate` for disambiguation rules). (see `CompatibilityCheckDetails`, `MultipleCandidatesDetails`)

We should not stray from these terms, as they have strict meaning in how they interact with the rules

Reply:
> Using ImmutableAttributes here is a matter of asserting the memory safety of the code over any other reason. Its nice implementing this code knowing that nobody will go and mutate any of these containers after the fact. We can even go ahead and cache some of the results here without fear of the container being mutated after the fact.

I really don't follow this reasoning at all.  A method shouldn't/can't care about what happens to its arguments after it returns.  If the method needs to ensure memory safety during its operation, immutable types are one option, yes, but it could also use non-mutable types and appropriately lock on something.

The signature of the method suggests it does something that only applies to the immutable types.  This is misleading, and makes it harder to understand the purpose of the method and the class.

Reply:
Dependency management uses immutable attributes consistently and pervasively. I believe asserting we have immutable types at this stage in resolution is a net positive in the safety of the resolution engine. 

I will be leaving these as `ImmutableAttributes`.